### PR TITLE
Migrate to typescript

### LINKED ISSUE #1: Migrate to TypeScript
Once we migrate to TypeScript (this primarily depends on migrating roboter to TypeScript), we can use this implementation:

```typescript
import humanizeString from 'humanize-string';

export interface ICustomError extends Error {
  name: string;
  code: string;
  message: string;
  cause?: Error;
}

type ErrorConstructors<T> = {
  [key in keyof T]: new(message?: string, cause?: Error) => ICustomError
};

const defekt = function <T extends {
  [ key: string ]: { code?: string };
}> (errorDefinitions: T): ErrorConstructors<T> {
  const errors: Partial<ErrorConstructors<T>> = {};

  /* eslint-disable guard-for-in */
  for (const errorName in errorDefinitions) {
    const errorDefinition = errorDefinitions[errorName];

    if (!errorDefinition) {
      continue;
    }

    const { code = `E${errorName.toUpperCase()}` } = errorDefinition;

    errors[errorName] = class extends Error implements ICustomError {
      public name: string;

      public code: string;

      public message: string;

      public cause?: Error;

      /* eslint-disable no-loop-func */
      public constructor (message = `${humanizeString(errorName)}.`, cause?: Error) {
        super();

        this.name = errorName;
        this.code = code;
        this.message = message;
        this.cause = cause;
      }
      /* eslint-enable no-loop-func */
    };
  }
  /* eslint-enable guard-for-in */

  return errors as ErrorConstructors<T>;
};

export default defekt;
```

### PR DESCRIPTION
**Pull request by yeldiRium**
Migrate the implementation to TypeScript. This contains breaking changes because we changed the configuration to better support TypeScript assistance.

Also adds CircleCI and Dependabot configs, as well as a status table in the readme.

This closes #1.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -15,17 +25,23 @@ $ npm install defekt
...
 ```javascript
-const errors = defekt([
-  'ArgumentNull',
-  'InvalidOperation',
-  '...'
-]);
+const errors = defekt({
+  ArgumentNull: {},
+  InvalidOperation: {},
+  // name: config
```
Comment (goloroden) [inline README.md]: This IMHO is misleading, as this suggests that there exists a key `name` with some special meaning … can we simply put

```
// ...
```

into this line?

File: README.md [hunk 2]
```diff
@@ -63,33 +79,20 @@ The custom errors follow the same rules as the built-in ones, i.e. they have a `
 
 By default, each custom error uses its uppercased name with an `E` prefix as error code. E.g., an `InvalidOperation` error uses `EINVALIDOPERATION` as its code.
 
-From time to time, you may want to provide custom error codes. For that specify an object with a `name` and a `code` property instead of only providing the error name:
+From time to time, you may want to provide custom error codes. For that specify a `code` property on the configuration objects:
```
Comment (goloroden) [inline README.md]: I'd use singular for `configuration object` here, as you don't have to provide it always (it's not or all-or-nothing decision).

File: README.md [hunk 3]
```diff
@@ -63,33 +79,20 @@ The custom errors follow the same rules as the built-in ones, i.e. they have a `
...
 ```javascript
-const errors = defekt([
-  { name: 'ArgumentNull', code: 'ARGNULL' },
-  { name: 'InvalidOperation', code: 'INVALOP' },
-  { name: '...', code: '...' }
-]);
+const errors = defekt({
+  ArgumentNull: { code: 'ARGNULL' },
+  InvalidOperation: { code: 'INVALOP' },
+  ...: { code: '...' }
```
Comment (goloroden) [inline README.md]: Again, here I'd use

```
// ...
```

File: README.md (hunk 1 above)
Comment (yeldiRium) [reply README.md]: I'd like to have some explanation of the object's structure here. Do you think the new version with a sentence more is better?

Comment (goloroden) [reply README.md]: Yes, I like it a lot :-)

**Pull request merged**